ibis/udf/vectorized.py

```diff
@@ -30,6 +36,34 @@ def __init__(self, func, func_type, input_type, output_type):
         self.func_type = func_type
         self.input_type = list(map(dt.dtype, input_type))
         self.output_type = dt.dtype(output_type)
+        self.coercion_fn = self._get_coercion_function()
+
+    def _get_coercion_function(self):
+        """Return the appropriate function to coerce the result of the UDF,
+        according to the func type and output type of the UDF."""
+        if isinstance(self.output_type, dt.Struct):
+            # Case 1: Struct output, non-reduction UDF -> coerce to DataFrame
+            if (
+                self.func_type is ElementWiseVectorizedUDF
+                or self.func_type is AnalyticVectorizedUDF
+            ):
+                return coerce_to_dataframe
+            else:
+                # Case 2: Struct output, reduction UDF -> coerce to tuple
+                return coerce_to_tuple
+        # Case 3: Vector output, non-reduction UDF -> coerce to Series
+        elif (
+            self.func_type is ElementWiseVectorizedUDF
+            or self.func_type is AnalyticVectorizedUDF
+        ):
+            return coerce_to_series
+        # Case 4: Array output type, reduction UDF -> coerce to np.ndarray
+        elif isinstance(self.output_type, dt.Array):
+            return coerce_to_np_array
```

Comment:
Do we have test to cover this?

Reply:
Yes this test hits it: https://github.com/ibis-project/ibis/blob/master/ibis/backends/dask/tests/test_udf.py#L487
I can add a more general one for all backends in test_vectorized_udf.py